Window Managers/Stump Config.org
#+title: Stump Config
#+author: Ian Kollipara
#+date: May 31, 2022
#+property: header-args :tangle ~/.config/stumpwm/config

* Table of Contents :toc:
- [[#introduction][Introduction]]
- [[#initial-setup][Initial Setup]]
  - [[#imports][Imports]]
  - [[#misc][Misc]]
- [[#startup-applications][Startup Applications]]
- [[#keybinds][Keybinds]]
  - [[#prefix-key][Prefix Key]]
  - [[#window-management][Window Management]]
  - [[#group-management][Group Management]]
  - [[#applications][Applications]]
- [[#groups][Groups]]
- [[#mode-line][Mode-line]]
  - [[#wifi][Wifi]]
  - [[#volume][Volume]]
  - [[#battery][Battery]]
  - [[#group-format-configuration][Group Format Configuration]]
  - [[#time-and-date][Time and Date]]
  - [[#full-mode-line-format][Full Mode-line Format]]
- [[#commands][Commands]]
- [[#gaps][Gaps]]
- [[#fonts][Fonts]]
- [[#message-and-input-bar][Message and Input Bar]]

* Introduction
This is my *StumpWM* Config. Much of it is hand-crafted, although some is pulled from [[https://github.com/Gavinok/stump-conf/blob/main/config][Gavin Freeborn's Config]]. The theme of the system is [[https://github.com/catppuccin][Catppuccin]]. I use Emacs for a majority of tasks, but not all of them. Emacs is used for the terminal emulator via =vterm=, as the primary editor, and as my email client via =mu4e=. This config facilitates much of that for me.

* Initial Setup
For the config to work, a few things have to be setup before I continue.
#+begin_src lisp
(in-package :stumpwm)
#+end_src

** Imports
#+begin_src lisp
(ql:quickload :slynk)
(ql:quickload :clx-truetype)
(ql:quickload :uiop)
(ql:quickload :str)
(ql:quickload :arrows)
(load-module "swm-gaps")
(load-module "ttf-fonts")
#+end_src

** Misc
I really don't like the "X" mouse cursor, so I change it to be consistent.
#+begin_src lisp
(stumpwm:run-shell-command "xsetroot -cursor_name left_ptr")
#+end_src

Another issue is font rendering, I use [[https://www.programmingfonts.org/#firacode][Fira Code]] for my font, which is modified as a Nerd Font, so I have icons I can use. But to do this, I have to update the cache fonts.
#+begin_src lisp
(xft:cache-fonts)
(clx-truetype:cache-fonts)
#+end_src

* Startup Applications
Stump provides a good mechanism for running autostart applications. This comes in the form of the variable ~*initializing*~. I autostart Dropbox, Emacs Daemon, Nitrogren, Picom, ... However, in the case of dropbox I don't start it unless I'm using Arch. For Fedora I install the dropbox package from their website.
#+begin_src lisp
(when *initializing*
  (when (eq (run-shell-command "uname -n" t) "fedora")
   (run-shell-command "~/.dropbox-dist/dropboxd"))
  (run-shell-command "picom")
  (run-shell-command "nitrogen --restore")
  (run-shell-command "emacs --daemon"))
#+end_src

* Keybinds
For my keybinds I use a helper function that just wraps the normal StumpWM keybind behavior. Its purely for conciseness.
#+begin_src lisp
(defun ik/keybind (map key cmd)
  (define-key map (kbd key) cmd))
#+end_src

** Prefix Key
Since I'm a big fan of Doom Emacs, I use =s-space= for my prefix. This is really close to doom's =space= prefix, and thus is less context switching for me.
#+begin_src lisp
(set-prefix-key
 (kbd "s-space"))
#+end_src

** Window Management
For moving, focusing, resizing, and splitting windows I use vim-like keybinds. All of these are bound at the top level.
#+begin_src lisp
(ik/keybind *top-map*  "s-h" "move-focus left")
(ik/keybind *top-map*  "s-l" "move-focus right")
(ik/keybind *top-map*  "s-k" "move-focus up")
(ik/keybind *top-map*  "s-j" "move-focus down")

(ik/keybind *top-map*  "s-M-h" "move-window left")
(ik/keybind *top-map*  "s-M-l" "move-window right")
(ik/keybind *top-map*  "s-M-j" "move-window down")
(ik/keybind *top-map*  "s-M-k" "move-window up")

(ik/keybind *root-map* "R" "iresize")
(ik/keybind *root-map* "r" "remove")
(ik/keybind *root-map* "s" "vsplit-and-focus")
(ik/keybind *root-map* "S" "hsplit-and-focus")
#+end_src

** Group Management
I also use pretty normal keybinds for group management. My previous window manager was Qtile, so I incorporated many features from that that I like. In particular this is the use of Super+$n$ for choosing a group, and s-, s-. for moving between groups next to each other.
#+begin_src lisp
(ik/keybind *top-map* "s-1" "gselect 1")
(ik/keybind *top-map* "s-2" "gselect 2")
(ik/keybind *top-map* "s-3" "gselect 3")
(ik/keybind *top-map* "s-4" "gselect 4")
(ik/keybind *top-map* "s-5" "gselect 5")
(ik/keybind *top-map* "s-6" "gselect 6")
(ik/keybind *top-map* "s-7" "gselect 7")
(ik/keybind *top-map* "s-8" "gselect 8")
(ik/keybind *top-map* "s-9" "gselect 9")

(ik/keybind *top-map* "s-." "gnext")
(ik/keybind *top-map* "s-," "gprev")
#+end_src

** Applications
Certain applications are nice to have bound to keys. Some examples are emacs, vterm, brave, etc.
#+begin_src lisp
(ik/keybind *top-map* "s-w" "brave")
(ik/keybind *top-map* "s-RET" "term")
(ik/keybind *root-map* "e" "exec emacsclient -c -a 'emacs'")
(ik/keybind *root-map* "C-e" "exec emacsclient -c -a 'emacs'")
#+end_src

* Groups
For my groups I have 9, but they are more or less general purpose. Some have a specific name, but they're really all just static workspaces.
#+begin_src lisp

(defvar *workspaces* '("Home"
                       "Web"
                       "Emacs"
                       "Office"
                       "Mail"
                       "Read"
                       "Note"
                       "Ref"))

(grename (first *workspaces*))
(loop for w in (rest *workspaces*)
      do (gnewbg w))
#+end_src

* Mode-line
For my mode-line I have a few extra modules that I add. These are for battery, wifi, and volume. There are some contrib modules for these, but they aren't suited for what I want.

** Wifi
For wifi I'm using NetworkManager, so I have a really simple script that queries for the network name. It also handles displaying different icons based on connectivity.
#+begin_src lisp
(defun wifi-string ()
    (str:remove-punctuation (run-shell-command "nmcli device | grep '[^dis]connected' | awk '{print $4}'" t)))
(defun wifi ()
    (if (equal (wifi-string) "")
        "睊 Disconnected"
        (str:concat "直 " (wifi-string))))
#+end_src

** Volume
For volume I wanted to show icons for different levels, as well as the state of volume. This requires more than just one function.
#+begin_src lisp
(defparameter *vol-icon-alist* '(
                                 (0 . "ﱝ")
                                 (10 . "奄")
                                 (50 . "奔")
                                 (100 . "墳")))


(defun vol-icon (vol-percent)
  "Get the corresponding icon given the volume percent"
  (cdr (assoc vol-percent *vol-icon-alist* :test #'<=)))

(defun vol-on-p (vol-line)
  "Predicate to tell if volume is on"
  (arrows:->> vol-line
              (str:split " ")
              sixth
              str:remove-punctuation
              (equal "on")))

(defun vol-percent-as-integer (vol-line)
  "Parse the line and return the volume string"
  (arrows:->> vol-line
              (str:split " ")
              fifth
              str:remove-punctuation
              parse-integer))

(defun vol ()
  "Create the Volume String"
  (let* ((vol-line (str:remove-punctuation (run-shell-command "amixer sget 'Master' | awk '/Left:/ {print}'" t)))
         (vol-level (vol-percent-as-integer vol-line)))
    (if (vol-on-p vol-line)
        (arrows:-> vol-level
                    vol-icon
                    (str:concat (format nil " ~d%" vol-level)))
        (arrows:-> 0
                   vol-icon
                   (str:concat " Muted")))))

#+end_src

** Battery
Battery isn't the hardest to figure out. I borrow quite a bit of this from [[https://github.com/solbloch/stumpwm-configs/blob/master/battery.lisp][solbloch's custom config]].
#+begin_src lisp
(defvar *battery-discharging-icon-alist* '((10 . "")
                                           (20 . "")
                                           (30 . "")
                                           (40 . "")
                                           (50 . "")
                                           (60 . "")
                                           (70 . "")
                                           (80 . "")
                                           (90 . "")
                                           (100 . "")))

(defvar *battery-charging-icon-alist* '((20 . "")
                                        (30 . "")
                                        (40 . "")
                                        (60 . "")
                                        (80 . "")
                                        (90 . "")
                                        (100 . "")))

;; Modified from solbloch
(defun battery ()
  (first (remove-if-not #'(lambda (supply)
                           (search "BAT" (namestring supply)))
                       (uiop:subdirectories "/sys/class/power_supply/"))))

;; Taken from solbach
(defun battery-alist (battery)
  (with-open-file (stream (str:concat (namestring battery) "uevent"))
    (loop for line = (read-line stream nil)
         while line
         collect (let ((line-split (str:split #\= line)))
                   (cons (car line-split) (cadr line-split))))))

(defun battery-discharging-p (bat)
  "Check if the battery is discharging or not"
  (equal "Discharging" (cdr( assoc "POWER_SUPPLY_STATUS" bat :test #'string=))))

(defun battery-state ()
  "Get the Battery String"
  (let* ((bat (battery-alist (battery)))
         (percent (parse-integer (cdr (assoc "POWER_SUPPLY_CAPACITY" bat :test #'string=)))))
    (if (battery-discharging-p bat)
        (str:concat (cdr (assoc percent *battery-discharging-icon-alist* :test #'<=)) (format nil " ~d%" percent))
        (str:concat (cdr (assoc percent *battery-charging-icon-alist* :test #'<=)) (format nil " ~d%" percent)))))
#+end_src

** Group Format Configuration
For displaying groups I use a few custom formatting elements. These wrap the active group with brackets ([]).
#+begin_src lisp
(defun current-group-left-bracket (group)
  "If current group, return left brack '['"
  (if (eq group (current-group))
      "["
      ""))

(defun current-group-right-bracket (group)
  "If current group, return right brack ']'"
  (if (eq group (current-group))
      "]"
      ""))

(defun current-group-font-color-start (group)
  (if (eq group (stumpwm:current-group))
      "^(:fg \"#f5a97f\")"
      ""))

(defun current-group-font-color-end (group)
  (if (eq group (stumpwm:current-group))
      "^(:fg \"#cad3f5\")"
      ""))

(setf *group-formatters*
      (append *group-formatters* '((#\[ current-group-left-bracket)
                                   (#\] current-group-right-bracket)
                                   (#\C current-group-font-color-start)
                                   (#\c current-group-font-color-start))))

(setf *group-format* "%[%C%t%c%]")
#+end_src

** Time and Date
For time and date I use a pretty standard layout.
#+begin_src lisp
(setf *time-modeline-string* "^(:bg \"#a6da95\") %I:%M %p ^(:bg \"#ed8796	\")  %a %b %e, %Y")
#+end_src

** Full Mode-line Format
Given all this configuration, here is the final mode-line setup.
#+begin_src lisp
(setf *mode-line-background-color* "#1e1e2e")
(setf *screen-mode-line-format*
      (list "^B^(:fg \"#cdd6f4\")%g"
            "^>"
            "^B%d | "
            "^B"
            '(:eval (vol))
            " | "
            "^B"
            '(:eval (wifi))
            " | "
            "^B"
            '(:eval (battery-state))))

;; This is a little helper to remove highlighting of the current group.
(setf *mode-line-highlight-template* "~A")
#+end_src

* Commands
I use quite a few commands, some of which are designed to work for both Fedora and Arch Linux, as I use both systems quite a bit.
#+begin_src lisp
(defcommand brave () ()
  "Launch Brave Browser"
  (if (eq (str:remove-punctuation (run-shell-command "uname -n" t)) "fedora")
      (run-shell-command "brave-browser")
      (run-shell-command "brave")))

(defcommand emacs () ()
  "Run Emac from the daemon"
  (run-shell-command "emacsclient -c -a 'emacs'"))

(defcommand hsplit-and-focus () ()
  "Split the view horizontally and focus on the new window"
  (hsplit)
  (move-focus :left))

(defcommand vsplit-and-focus () ()
  "Split the view vertically and focus on the new window"
  (vsplit)
  (move-focus :up))

(defcommand obsidian () ()
  "Run Obsidian"
  (run-shell-command "flatpak run md.obsidian.Obsidian"))

(defcommand zotero () ()
  "Run Zotero"
  (run-shell-command "flatpak run org.zotero.Zotero"))

(defcommand only-office () ()
  "Run OnlyOffice"
  (run-shell-command "flatpak run org.onlyoffice.desktopeditors"))

(defparameter *stumpwm-slynk-session* nil)

(defcommand start-slynk () ()
  "Start Slynk for use in Emacs"
  (handler-case
      (defparameter *stumpwm-slynk-session*
        (slynk:create-server
         :dont-close t
         :port slynk::default-server-port))
    (error (c)
      (format *error-output* "Error starting slynk: ~a~%" c))))

(defcommand restart-slynk () ()
  "Restart the Slynk server if any issues arise"
  (slynk:stop-server slynk::default-server-port)
  (slynk:create-server :port slynk::default-server-port :dont-close t))

(defcommand select-wallpaper () ()
      (run-shell-command "nitrogen ~/Pictures/Wallpapers/"))

(defcommand term () ()
  (run-shell-command "emacsclient -c -a 'emacs' --eval '(+vterm/here \"~\")'"))
#+end_src

* Gaps
I like having gaps in my windows. I know they're useless, but I love their style.
#+begin_src lisp
(setf swm-gaps:*inner-gaps-size* 5)
(swm-gaps:toggle-gaps-on)
#+end_src
* Fonts
For my font, as I stated earlier, I'm using Fira Code Nerd Font. This is where I initialize that.
#+begin_src lisp
(set-font (make-instance 'xft:font :family "FiraCode Nerd Font" :subfamily "Regular" :size 13))
(mode-line)
#+end_src
* Message and Input Bar
For the message and input bar, I move it to the center of the screen, so it's easier to see and work with.
#+begin_src lisp
(setf *message-window-gravity* :center
      *message-window-input-gravity* :center
      *input-window-gravity* :center)
#+end_src


doom/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets. It is optional.
(setq user-full-name "Ian Kollipara"
      user-mail-address "ian.kollipara@cune.org")

(setq doom-font (font-spec :family "FiraCode Nerd Font" :size 20)
      doom-variable-pitch-font (font-spec :family "Cantarell" :size 20)
      doom-big-font (font-spec :family "Cantarell" :size 30))

;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-unicode-font' -- for unicode glyphs
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:
;;
;;(setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
;;      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
;;
;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq doom-theme 'doom-palenight)

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type t)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/Dropbox/Org/")


;; Whenever you reconfigure a package, make sure to wrap your config in an
;; `after!' block, otherwise Doom's defaults may override your settings. E.g.
;;
;;   (after! PACKAGE
;;     (setq x y))
;;
;; The exceptions to this rule:
;;
;;   - Setting file/directory variables (like `org-directory')
;;   - Setting variables which explicitly tell you to set them before their
;;     package is loaded (see 'C-h v VARIABLE' to look up their documentation).
;;   - Setting doom variables (which start with 'doom-' or '+').
;;
;; Here are some additional functions/macros that will help you configure Doom.
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;; Alternatively, use `C-h o' to look up a symbol (functions, variables, faces,
;; etc).
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how

(defvar ik/frame-transparency '(90 . 90))

(set-frame-parameter (selected-frame) 'alpha ik/frame-transparency)
(add-to-list 'default-frame-alist `(alpha . ,ik/frame-transparency))


(require 'smtpmail)
(set-email-account! "cune.org"
        '((mu4e-sent-folder . "/[Gmail].Sent Mail")
          (mu4e-trash-folder . "/[Gmail].Trash")
          (mu4e-refile-folder . "/[Gmail].All Mail")
          (mu4e-starred-folder . "/[Gmail].Starred")
          (smtpmail-smtp-user . "ian.kollipara@cune.org")
          (smtpmail-smtp-server . "smtp.gmail.com")
          (smtpmail-smtp-service . 587)
          (mu4e-compose-signature . "Sincerely,\nIan"))
        t)

(after! mu4e
  (setq +mu4e-gmail-accounts '(("ian.kollipara@cune.org" . "/cune"))
        mu4e-index-cleanup t
        mu4e-index-lazy-check t))

(setq! bibtex-completion-bibliography '("/home/ikollipara/Dropbox/Zettlekasten/My Library.bib")
       citar-bibliography '("/home/ikollipara/Dropbox/Zettlekasten/My Library.bib")
       reftex-default-bibliograpy "/home/ikollipara/Dropbox/Zettlekasten/My Library.bib")

(add-hook! 'python-mode-hook (modify-syntax-entry ?_ "w"))

(after! org
  (setq org-hide-emphasis-markers t))

(setq +latex-viewers '(pdf-tools))

(after! org-journal
  (setq org-journal-dir "/home/ikollipara/Dropbox/REU/UTD/Research/Research Journal"
        org-journal-date-format "%x - %A"))


(after! smudge
 (setq smudge-oauth2-client-id "08be6f422d0d486eadbe9207f0eda57c")
 (setq smudge-oauth2-client-secret "f4f560fce2ca42e0bef597a3e4451f96")
 (setq smudge-transport 'connect)
 (map! :leader
       (:prefix ("S" . "Smudge")
        :desc "Toggle Shuffle" "s" #'smudge-controller-toggle-shuffle
        :desc "Toggle Repeat" "r" #'smudge-controller-toggle-repeat
        :desc "Play/Pause" "p" #'smudge-controller-toggle-play
        :desc "Next Track" ">" #'smudge-controller-next-track
        :desc "Prev Track" "<" #'smudge-controller-previous-track
        (:prefix ("P" . "playlist")
         :desc "My Playlists" "m" #'smudge-my-playlists
         :desc "Featured Playlists" "f" #'smudge-featured-playlists
         :desc "Search Playlists" "s" #'smudge-playlist-search)
        (:prefix ("t" . "track")
         :desc "Recently Played" "r" #'smudge-recently-played
         :desc "Search Track" "s" #'smudge-track-search))))


(add-hook! (org-mode-hook latex-mode-hook) #'citar-refresh)

(after! doom-modeline
  (setq display-time-day-and-date t)
  (display-battery-mode 1)
  (display-time-mode 1))

(defun wifi-string ()
  "Get the current network"
  (let ((wifi (shell-command-to-string "nmcli d status | awk '/[^dis]connected/ {print $4}'")))
    (if (equal wifi "")
        "睊 Disconnected"
        (concat "直 " (substring wifi 0 -1)))))

(setq-default header-line-format '(
                                   (" ")
                                   (:eval (capitalize (+workspace-current-name)))
                                   (" | ")
                                   (:eval (wifi-string))))

(setq org-noter-notes-search-path '("~/Dropbox/REU/UTD/Research/Actor Model Design for MCPSs through Rebeca/Notes/"))

(add-to-list '+lookup-provider-url-alist '("Google Scholar" "https://scholar.google.com/scholar?q=%s"))

(map! :map LaTeX-mode-map
      :localleader
      :nvm "@" 'citar-insert-citation
      :nvm "s" 'LaTeX-section
      :nvm "e" 'LaTeX-environment
      :i "C-j" 'LaTeX-insert-item)

(map! :leader
      :nvm "/" 'consult-line)
